conan/tools/gnu/makedeps.py

```diff
@@ -0,0 +1,588 @@
+            formatted_dirs = _get_formatted_dirs(cppinfo_value, self._root, _makefy(self._name))
+            if formatted_dirs:
+                self._makeinfo.dirs.append(var)
+                var = var.replace("dirs", "_dirs")
+                formatted_dirs = self._rootify(self._root, self._dep.ref.name, cppinfo_value)
+                dirs[var] = [_conan_prefix_flag(flag) + it for it in formatted_dirs]
+        return dirs
+
+    def _rootify(self, root: str, root_id: str, path_list: list) -> list:
+        """
+        Replaces component folder path by its root node folder path in case they match
+        :param root: root folder path for component's father
+        :param root_id: component's dependency name
+        :param path_list: folder list available in the component
+        :return: A formatted folder list, solving root folder path as prefix
+        """
+        root_len = len(root)
+        root_with_sep = root + os.sep
+        root_var_ref = f"$(CONAN_ROOT_{_makefy(root_id)})"
+        return [root_var_ref + path[root_len:].replace("\\", "/") if path.startswith(root_with_sep) else path for path in path_list]
+
+    def _get_component_flags(self) -> dict:
+        """
+        List common variables from cpp_info and format them to be used in the makefile
+        :return: A dictionary with regular flag/variable name and its formatted value with prefix
+        """
+        flags = {}
+        for var, prefix_var in _common_cppinfo_variables().items():
+            cppinfo_value = getattr(self._comp, var)
+            if not cppinfo_value:
+                continue
+            if "flags" in var:
+                cppinfo_value = [var.replace('"', '\\"') for var in cppinfo_value]
+            if cppinfo_value:
+                flags[var.upper()] = [_conan_prefix_flag(prefix_var) + it for it in cppinfo_value]
+                self._makeinfo.flags.append(var)
+        return flags
+
+    def generate(self) -> str:
+        """
+        Process component cpp_info variables and generate its Makefile content
+        :return: Component Makefile content
+        """
+        dirs = self._get_component_dirs()
+        flags = self._get_component_flags()
+        comp_content_gen = DepComponentContentGenerator(self._dep, self._name, dirs, flags)
+        comp_content = comp_content_gen.content()
+        return comp_content
+
+
+class DepGenerator:
+    """
+    Process a dependency cpp_info variables and generate its Makefile content
+    """
+
+    def __init__(self, dependency: object):
+        self._dep = dependency
+        self._info = MakeInfo(self._dep.ref.name, [], [])
+
+    def make_global_info(self) -> MakeInfo:
+        """
+        Remove duplicated folders and flags from current MakeInfo
+        :return: Dependency folder and flags
+        """
+        self._info.dirs = list(set(self._info.dirs))
+        self._info.flags = list(set(self._info.flags))
+        return self._info
```

Comment:
This way of removing duplicates can and probably will change the order of directories and flags, which is likely an undesired behavior. The defined order by users and topology of the graph must be respected.